interop/nkic/gather.c

```
@@ -5,51 +5,911 @@ Released under Apache 2.0 license as described in the file LICENSE.
+      return true;
+  }
+  return false;
+}
+
+static bool have_global(struct state *st, const char *name) {
+  for (struct Python_Keyword_List *l = st->globals; l; l = l->next) {
+    if (strcmp(l->keyword->id, name) == 0)
+      return true;
+  }
+  return false;
+}
+
+// Copy a Python string to our memory region.
+// Note: we disallow embedded NULLs (which can only show up in literals).
+static char* py_strdup(struct state *st, PyObject *obj) {
+  Py_ssize_t sz = 0;
+  const char *s = PyUnicode_AsUTF8AndSize(obj, &sz);
+  if (!s || sz <= 0)
+    return NULL;
+
+  if (memchr(s, 0, sz)) {
+    syntax_error(st, "embedded NULL in string");
+    return NULL;
+  }
+
+  return region_strdup(st->region, s);
+}
+
+// Construct a path name from two strings.
+static char* path_name(struct state *st, const char *m, const char *x) {
+  if (!m || !x)
+    return NULL;
+
+  size_t m_sz = strlen(m);
+  size_t x_sz = strlen(x);
+
+  char *name = region_alloc(st->region, m_sz + x_sz + 2);
+  if (!name)
+    return NULL;
+
+  memcpy(name, m, m_sz);
+  name[m_sz] = '.';
+  memcpy(name + m_sz + 1, x, x_sz);
+  name[m_sz + x_sz + 1] = 0;
+  return name;
+}
+
+// Construct a path name from two Python strings (in memory region)
+static inline char* py_path_name(struct state *st, PyObject *m, PyObject *x) {
+  if (!m || !x)
+    return NULL;
+  return path_name(st, PyUnicode_AsUTF8(m), PyUnicode_AsUTF8(x));
+}
+
+// Construct full name of python function (in memory region)
+static char* py_fun_name(struct state *st, PyObject *f) {
+  PyObject *module = PyObject_GetAttrString(f, "__module__");
+  PyObject *name = PyObject_GetAttrString(f, "__name__");
+  char *f_name = py_path_name(st, module, name);
+
+  Py_XDECREF(module);
+  Py_XDECREF(name);
+  return f_name;
+}
+
+// Add a new function to the work-list (if necessary)
+static void add_work(struct state *st, PyObject *f) {
+  if (!PyFunction_Check(f))
+    return;
+
+  char *name = py_fun_name(st, f);
+  if (!name)
+    return;
```

add_work() has no way to report whether an error occurred.

Don't we want to stop processing immediately, vs ignoring the error and then something way later fails mysteriously?